File: behavioral/chain-of-responsibility/chain-of-responsibility.java
public class ChainOfResponsibilityPattern {
    public enum Dishes {
        breakfast,
        dinner,
        supper
    }

    public static abstract class Cat {
        protected String name;
        protected Cat successor;

        public Cat(String name, Cat successor) {
            this.name = name;
            this.successor = successor;
        }

        protected abstract String prepareDish();
        protected abstract Dishes dish();

        public void handleRequest(Dishes dish) {
            if (dish() == dish) {
                String result = String.format("%s: %s", name, prepareDish());
                System.out.println(result);
            } else if (successor != null) {
                successor.handleRequest(dish);
            }
        }
    }

    public static class Siamese extends Cat {
        public Siamese(String name, Cat successor) {
            super(name, successor);
        }

        protected Dishes dish() {
            return Dishes.breakfast;
        }

        protected String prepareDish() {
            return "Breakfast is ready! Bon Appetit!";
        }
    }

    public static class MaineCoon extends Cat {
        public MaineCoon(String name, Cat successor) {
            super(name, successor);
        }

        protected Dishes dish() {
            return Dishes.dinner;
        }

        protected String prepareDish() {
            return "There is a dinner! Let's begin!";
        }
    }

    public static class BritishShorthair extends Cat {
        public BritishShorthair(String name, Cat successor) {
            super(name, successor);
        }

        protected Dishes dish() {
            return Dishes.supper;
        }

        protected String prepareDish() {
            return "Time for supper! Yeah!";
        }
    }

    public static void main(String[] args) {
        Cat britishShorthair = new BritishShorthair("Oliver", null);
        Cat maineCoon = new MaineCoon("Oscar", britishShorthair);
        Cat siamese = new Siamese("Simba", maineCoon);

        siamese.handleRequest(Dishes.breakfast);
        siamese.handleRequest(Dishes.dinner);
        siamese.handleRequest(Dishes.supper);
    }
}

File: behavioral/chain-of-responsibility/chain-of-responsibility.ts
enum Dishes {
    breakfast,
    dinner,
    supper
}

abstract class Cat {
    protected _name: string;
    protected _successor: Cat;
    protected abstract _dish: Dishes;

    constructor(name: string, successor?: Cat) {
        this._name = name;
        this._successor = successor;
    }

    protected abstract _prepareDish(): string;

    public handleRequest(dish: Dishes): void {
        if (this._dish === dish) {
            console.log(`${this._name}: ${this._prepareDish()}`);
        } else if (this._successor) {
            this._successor.handleRequest(dish);
        }
    }
}

class Siamese extends Cat {
    protected _dish: Dishes = Dishes.breakfast;

    protected _prepareDish(): string {
        return "Breakfast is ready! Bon Appetit!";
    }
}

class MaineCoon extends Cat {
    protected _dish: Dishes = Dishes.dinner;

    protected _prepareDish(): string {
        return "There is a dinner! Let's begin!";
    }
}

class BritishShorthair extends Cat {
    protected _dish: Dishes = Dishes.supper;

    protected _prepareDish(): string {
        return "Time for supper! Yeah!";
    }
}

let britishShorthair: Cat = new BritishShorthair("Oliver");
let maineCoon: Cat = new MaineCoon("Oscar", britishShorthair);
let siamese: Cat = new Siamese("Simba", maineCoon);

siamese.handleRequest(Dishes.breakfast);
siamese.handleRequest(Dishes.dinner);
siamese.handleRequest(Dishes.supper);
